### PR TITLE
Add support for specifying a predicate to find a specific element

### LINKED ISSUE #37: Allow specifying a predicate function to only resolve on a more specific element
The current behavior of `elementReady` is great, but sometimes there are multiple items matching the selector. Currently, this means that the first matching element will be returned. However, there are many use-cases where other ones are wanted, and a more specific CSS selector cannot be used. For example:
```html
<ul id="country-list">
	<li>country a</li>
	...
	<li>wanted country</li>
	...
</ul>
```

Currently, you can't get the `<li>` with `wanted country`, where the entire `<ul>` (or `<li`s) is not present on the page from the get-go.

My suggestion is to add an **optional** `predicate` function to the `elementReady` options, such that you could solve the above scenario like so:
```js
const someSpecificCountry = await elementReady('#country-list > li', {
	predicate: item => !!item.textContent && /wanted country/i.test(item.textContent),
});
```

I already have a working implementation of this as a patch (using [`patch-package`](https://www.npmjs.com/package/patch-package)) and would be happy to submit a PR if this is an acceptable for the problem.

### PR DESCRIPTION
**Pull request by bengry**
fixes #37

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ export default function elementReady(selector, {
 	stopOnDomReady = true,
 	waitForChildren = true,
 	timeout = Number.POSITIVE_INFINITY,
+	predicate = () => true,
```
Comment (sindresorhus) [inline index.js:15]: It's wasteful to execute an empty function for each element.

File: index.d.ts [hunk 2]
```diff
@@ -32,6 +32,11 @@ export interface Options {
 	@default true
 	*/
 	readonly waitForChildren?: boolean;
+
+	/**
```
Comment (sindresorhus) [inline index.d.ts:36]: This needs a description (including mentioning its use-cases), and ideally a realistic code example.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
This will need readme updates and tests.

**Comment by sindresorhus** [comment]
It should also be added to `observeReadyElements`.

**Comment by bengry** [comment]
> This will need readme updates and tests.

I added this to the readme, but I can't run the tests locally since it seems like the `typescript` package is missing from `devDependencies`. Is that on purpose? I'm not sure how they pass in CI.

> It should also be added to `observeReadyElements`.

Done

**Comment by bengry** [comment]
On an unrelated note, is there a reason to maintain a separate `.d.ts` file, instead of re-writing the package in TypeScript? It's not large and a re-write should be simple enough.

**Comment by sindresorhus** [comment]
> On an unrelated note, is there a reason to maintain a separate .d.ts file, instead of re-writing the package in TypeScript?

Yes. I don't like TS. I only provide types for user convenience. (https://github.com/sindresorhus/ky/issues/321#issuecomment-765594679)

File: readme.md [hunk 3]
```diff
@@ -86,6 +86,13 @@ By default, `element-ready` guarantees the element and all of its children have
 
 By setting this to `false`, `element-ready` will resolve the promise as soon as it finds the requested selector, regardless of its content. This is ok if you're just checking if the element exists or if you want to read/change its attributes.
 
+##### predicate
+
+Type: `(element: HTMLElement) => boolean`\
+Default: `undefined`
+
+If defined, the predicate will be called for each element that matches the selector. If it returns `true`, the element will be returned.
```
Comment (sindresorhus) [inline readme.md:94]:
```suggestion
A predicate function will be called for each element that matches the selector. If it returns `true`, the element will be returned.
```

File: index.d.ts [hunk 4]
```diff
@@ -32,6 +32,25 @@ export interface Options {
...
+ 	If defined, the predicate will be called for each element that matches the selector. If it returns `true`, the element will be returned.
+
+	For example, if the content is dynamic or a selector can't be specific enough - you can check the `.textContent` of each element and only match the one that has the required text.
+
+	@example
```
Comment (sindresorhus) [inline index.d.ts:41]: The example is missing the JS part, like in your issue.

And this example also needs to be added to the readme.

Generally, I try to keep index.d.ts and readme in sync as much as possible.

File: index.js [hunk 5]
```diff
@@ -125,3 +127,13 @@ export function observeReadyElements(selector, {
 		},
 	};
 }
+
+function getMatchingElement({target, selector, predicate} = {}) {
+	if (predicate) {
+		const elements = target.querySelectorAll(selector);
+		const element = [...elements].find(element => predicate(element));
+		return element;
```
Comment (sindresorhus) [inline index.js:135]:
```suggestion
		return [...elements].find(element => predicate(element));
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 6]
```diff
@@ -32,6 +32,31 @@ export interface Options {
 	@default true
 	*/
 	readonly waitForChildren?: boolean;
+
+	/**
+ 	If defined, the predicate will be called for each element that matches the selector. If it returns `true`, the element will be returned.
```
Comment (sindresorhus) [inline index.d.ts:37]: Readme and index.d.ts should be in sync.

File: index.d.ts [hunk 7]
```diff
@@ -32,6 +32,31 @@ export interface Options {
...
+
+	For example, if the content is dynamic or a selector can't be specific enough - you can check the `.textContent` of each element and only match the one that has the required text.
```
Comment (sindresorhus) [inline index.d.ts:41]: Readme and index.d.ts should be in sync.

Comment (sindresorhus) [inline index.d.ts:41]:
```suggestion
	For example, if the content is dynamic or a selector cannot be specific enough, you could check `.textContent` of each element and only match the one that has the required text.
```

**Comment by sindresorhus** [comment]
You overwrote my changes: https://github.com/sindresorhus/element-ready/compare/8dc7b0c24363479246a6a2b4b3cf3d6da6f968c4..ed957e569374df606507811648b6f2730b983a2c

**Comment by sindresorhus** [comment]
Generally, don't force push in pull requests, it also makes it harder to review what changed.

**Comment by bengry** [comment]
> You overwrote my changes: [`8dc7b0c24363479246a6a2b4b3cf3d6da6f968c4..ed957e569374df606507811648b6f2730b983a2c` (compare)](https://github.com/sindresorhus/element-ready/compare/8dc7b0c24363479246a6a2b4b3cf3d6da6f968c4..ed957e569374df606507811648b6f2730b983a2c)

I've re-added them in add7485. If I understood correctly - with the last push all that was overridden was the missing `import elementReady from 'element-ready';` from the docs. Since there were both this and the suggestions (which I've implemented separately) in the diffs.

Please LMK if I missed anything.

**Comment by sindresorhus** [comment]
There were a few other changes too. Just look at the diff I commented.

**Comment by bengry** [comment]
> There were a few other changes too. Just look at the diff I commented.

Right, but there were also the [suggestion](https://github.com/sindresorhus/element-ready/pull/38#discussion_r721166796), so I think that together we're all aligned now: https://github.com/sindresorhus/element-ready/compare/8dc7b0c24363479246a6a2b4b3cf3d6da6f968c4..add7485

**Comment by sindresorhus** [comment]
I committed some small tweaks.

**Comment by sindresorhus** [comment]
The "check elements against a predicate" test is failing though.

**Comment by bengry** [comment]
> The "check elements against a predicate" test is failing though.

Fixed in cd1f75a. There were some errors due to some renaming I did while working on the tests, and using a non-existent property on the element in the assertion. Funnily enough, these errors would've been caught by TypeScript.

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**